rust/main/utils/run-locally/src/main.rs

```diff
@@ -510,49 +450,224 @@ fn main() -> ExitCode {
+        )
     } else {
+        relayer_env.arg("relayChains", "test1,test2,test3")
+    }
+}
+
+/// Kills relayer in State and respawns the relayer again
+fn restart_relayer(config: &Config, state: &mut State, rocks_db_dir: &TempDir) {
+    log!("Stopping relayer...");
+    let (child, _) = state.agents.get_mut("RLY").expect("No relayer agent found");
+    child.kill().expect("Failed to stop relayer");
+
+    log!("Restarting relayer...");
+    let relayer_env = create_relayer(config, rocks_db_dir);
+    state.push_agent(relayer_env.spawn("RLY", Some(&AGENT_LOGGING_DIR)));
+    log!("Restarted relayer...");
+}
+
+/// Check relayer restart behaviour is correct.
+/// So far, we only check if undelivered messages' statuses
+/// are correctly retrieved from the database
+fn relayer_restart_invariants_met() -> eyre::Result<bool> {
+    let log_file_path = AGENT_LOGGING_DIR.join("RLY-output.log");
+    let relayer_logfile = File::open(log_file_path).unwrap();
+
+    let line_filters = vec![RETRIEVED_MESSAGE_LOG, "CouldNotFetchMetadata"];
+
+    log!("Checking message statuses were retrieved from logs...");
+    let matched_logs = get_matching_lines(&relayer_logfile, vec![line_filters.clone()]);
+
+    let no_metadata_message_count = *matched_logs
+        .get(&line_filters)
+        .expect("Failed to get matched message count");
+    // These messages are never inserted into the merkle tree.
+    // So these messages will never be deliverable and will always
+    // be in a CouldNotFetchMetadata state.
+    // When the relayer restarts, these messages' statuses should be
+    // retrieved from the database with CouldNotFetchMetadata status.
+    if no_metadata_message_count < ZERO_MERKLE_INSERTION_KATHY_MESSAGES {
+        log!(
+            "No metadata message count is {}, expected {}",
+            no_metadata_message_count,
+            ZERO_MERKLE_INSERTION_KATHY_MESSAGES
+        );
+        return Ok(false);
+    }
+    assert_eq!(
+        no_metadata_message_count,
+        ZERO_MERKLE_INSERTION_KATHY_MESSAGES
+    );
+    Ok(true)
+}
+
+fn wait_for_condition<F1, F2, F3>(
+    config: &Config,
+    start_time: Instant,
+    condition_fn: F1,
+    loop_invariant_fn: F2,
+    mut shutdown_criteria_fn: F3,
+) -> bool
+where
+    F1: Fn() -> eyre::Result<bool>,
+    F2: Fn() -> bool,
+    F3: FnMut() -> bool,
+{
+    let loop_check_interval = Duration::from_secs(5);
+    while loop_invariant_fn() {
+        sleep(loop_check_interval);
+        if !config.ci_mode {
+            continue;
+        }
+        if condition_fn().unwrap_or(false) {
+            // end condition reached successfully
+            break;
+        }
+        if check_ci_timed_out(config.ci_mode_timeout, start_time) {
+            // we ran out of time
+            log!("CI timeout reached before invariants were met");
+            return false;
+        }
+        if shutdown_criteria_fn() {
+            return false;
```

Comment:
can you still do the `SHUTDOWN.store(true, Ordering::relaxed)` here? there may be some things running concurrently that would need to be indicated to shut down as well

Reply:
Would be kind of hard because both `wait_for_condition` depends on this `SHUTDOWN`.
If we did the SHUTDOWN here, then we need to reset it for the second `wait_for_condition`.
So you would need to do `Ctrl+C` twice to cancel the e2e locally.

Reply:
Imo a `shutdown_criteria_fn` being true should mean that the test essentially skips everything else, but does what it needs to gracefully shut down things. I think just adding a `SHUTDOWN.store(true, Ordering::relaxed)` would do this-- if we hit this for the first time we call `wait_for_condition`, we get back `test_passed` as false, and we end the test early with this
```
if !test_passed {
  log!("Failure occurred during E2E");
  return report_test_result(test_passed);
}
```

Which is the behavior we want imo

Reply:
ah I see